Feature/Optional Database Caching

Valentin-Vi opened the pull request:
## Description

Implements an optional in-memory caching layer for the Core database module using the builder pattern. The cache wraps the base repository interface with a cached repository decorator (decorator pattern) using Ristretto, a high-performance concurrent cache library. This is step 1 of the broader persistence initiative.

## Type of Change

- [ ] 🐛 Bug fix (fix/*)
- [x] ✨ New feature (feature/*)
- [ ] 🔧 Enhancement (enhancement/*)
- [ ] 🚑 Hotfix (hotfix/*)
- [ ] 🚀 Release (release/*)

## Related Issues

Closes #59

## Changes Made

- Implemented `CachedRepository` decorator that wraps base repository with transparent caching using Ristretto
- Added `CacheConfig` struct with Ristretto-specific settings (`NumCounters`, `MaxCost`, `BufferItems`, `DefaultTTL`, `DefaultCostOfItem`)
- Implemented `DatabaseBuilder` pattern with optional `WithCache()` method for flexible composition
- Added custom cache error types for better error handling
- Added `Cache` struct to config system (programmatic defaults, not YAML-configured)
- Maintained backwards compatibility: `NewDatabase()` function still works without changes
- Comprehensive test suite with ~42 tests and 10 benchmarks covering all components
- Added `github.com/dgraph-io/ristretto/v2` dependency (v2.3.0)

### New Files

- `internal/core/database/cache/cached_repository.go` - CachedRepository decorator using Ristretto
- `internal/core/database/cache/config.go` - CacheConfig struct and DefaultCacheConfig()
- `internal/core/database/cache/error/errors.go` - Custom cache error definitions
- `internal/core/database/cache/cached_repository_test.go` - Comprehensive tests including parity, integration, and benchmarks
- `internal/core/database/builder.go` - DatabaseBuilder pattern implementation
- `internal/core/database/builder_test.go` - Builder pattern tests

### Modified Files

- `internal/core/config/config.go` - Added Cache struct and GetCache() function
- `go.mod` - Added `github.com/dgraph-io/ristretto/v2` dependency

## Breaking Changes

- [ ] This PR introduces breaking changes
- [x] Migration guide provided (if yes)

### Migration Guide

**No breaking changes.** The cache is **optional and disabled by default**. Existing code using `NewDatabase()` continues to work without modification. This is a purely additive feature.

To enable caching, use the new builder pattern:

```go
// With cache enabled (using custom config)
db, err := database.NewDatabaseBuilder[MyEntity](ctx, "mydb").
    WithCache(cache.CacheConfig{
        NumCounters:       1e7,
        MaxCost:           100,
        BufferItems:       64,
        DefaultTTL:        5 * time.Minute,
        DefaultCostOfItem: 1,
    }, func(entity *MyEntity) string {
        return entity.ID.String()
    }).
    Build()

// With cache enabled (using defaults)
db, err := database.NewDatabaseBuilder[MyEntity](ctx, "mydb").
    WithCache(cache.DefaultCacheConfig(), extractKeyFunc).
    Build()

// Without cache (backwards compatible)
db, err := database.NewDatabase[MyEntity](ctx, "mydb")
```

## Checklist

- [x] Code follows Clean Architecture principles
- [x] Code follows project style guidelines
- [x] Self-review completed
- [x] Comments added for complex logic
- [ ] All documentation updated in `./docs/` folder
- [x] Tests pass locally (`make test`)
- [ ] Coverage requirements met (`make test-coverage`)
- [ ] No linting errors (`make lint`)
- [ ] Security scan passes (`make security`)
- [x] Branch name follows convention (enhancement/*, feature/*, fix/*, hotfix/*, release/*)

## Screenshots/Logs (if applicable)

### Test Coverage Summary

| Component | Tests | Coverage |
|-----------|-------|----------|
| CachedRepository | ~30 | Create, GetByID (hit/miss), Update, Delete, Exists, Count, Where, error paths |
| Parity Tests | 6 | Ensures cached repo behaves identically to base repo |
| Builder Pattern | 8 | Build with/without cache, chaining, backwards compatibility |
| Integration | 3 | CRUD flow, cache expiry, concurrent access |
| Benchmarks | 10 | Performance comparison: cached vs non-cached |

### Key Features

- **Optional**: Cache is disabled by default, zero overhead when not used
- **Thread-safe**: Uses Ristretto which is designed for high concurrency
- **Async writes**: Cache operations use goroutines for non-blocking performance
- **Cost-based eviction**: Ristretto uses intelligent admission/eviction policies
- **Automatic invalidation**: Cache updated on Create, invalidated on Update/Delete
- **TTL support**: Entries expire after configured duration
- **ExtractKey function**: Flexible key extraction for any entity type

Valentin-Vi committed:
Implemented optional thread-safe caching to database

Valentin-Vi committed:
Implemented optional thread-safe caching for database

Valentin-Vi committed:
Formatted

Valentin-Vi committed:
Wrote more tests to satisfy -gtl 90 percent test coverage

char2cs commented on internal/core/config/config.go:
```diff
@@ -142,6 +153,11 @@ func getDefaultConfig() *Config {
+				Enabled:         false,
+				DefaultTTL:      "5m",
+				CleanupInterval: "1m",
+			},
```
No dejaria la config desde la config general de Quiver. Pensa que cada modulo que implemente el cache va a leer esta config "default". 
La `config.yaml` esta pensada para integradores mas que nada, cosas que el usuario cambie o parametros nuestros internos, no se si se entiende.

valentin-villa replied:
No me quedo claro el incio, "No dejaria la config desde la config general de Quiver". Tampoco a que te referis con "integradorres".
Podrias explicarlo de otra manera?

char2cs replied:
Sii obvio. Cuando hablo de integradores, me refiero al código de Quiver que utilicé esta feature cuando esté mergeada (ej: event sourcing) 

Respecto a la config, acá estaríamos configurando la misma config para todos quienes implementan al caché. La config en quiver está pensada más para varios ítems:
- Reglas de negocio.
- Settings para mejorar performance.

Si me preguntas a mi, no tocaria este archivo de config, pasemos esa responsabilidad a integradores.

char2cs replied:
Revivo esto por aca, no creo que deberiamos dejar el cache configurable por config global de Quiver. Creo que el cache tiene que poder ser configurado por cada integrador.

char2cs commented on internal/core/database/builder.go:
```diff
@@ -0,0 +1,57 @@
+	"github.com/rabbytesoftware/quiver/internal/core/database/repository"
+)
+
+// DatabaseBuilder provides a builder pattern for creating database repositories with optional caching.
```
Trataría de sacar estos comments; si el código necesita comments, significa que hay que refactorizarlo. Sacando ese principio, el código está re prolijo, lo dejaría sin comments nada más

char2cs commented on internal/core/database/builder_test.go:
```diff
@@ -0,0 +1,200 @@
+
+// =============================================================================
+// Builder Pattern Tests
+// =============================================================================
```
Lo mismo aca que el anterior comment.

char2cs commented on internal/core/database/cache/config.go:
```diff
@@ -0,0 +1,34 @@
+func DefaultCacheConfig() CacheConfig {
+	return CacheConfig{
+		Enabled:         true,
+		DefaultTTL:      5 * time.Minute,
```
Magic numbers deberian ser constantes. 
Esta config default podria actuar como fallback si el integrador decide no pasar ninguna config.

Valentin-Vi committed:
Checks in TestGoCache_Get_UnmarshalError test were flawd. It intended to return an unmarshal error but it checked for if there was none. Updated logic to assert expected behaviour.
Checking for if:
1. found == true then error == nil
else
2. found == false then error != nil && error contains "unmarshal"

Valentin-Vi committed:
Ran  and [0;34mFormatting Go code...[0m
[0;32mCode formatted![0m  (mala mia chicos jajajaj)

char2cs commented on internal/core/database/cache/gocache.go:
```diff
@@ -0,0 +1,94 @@
+}
+
+// Set stores a value in the cache with the specified TTL.
+func (g *GoCache) Set(ctx context.Context, key string, value interface{}, ttl time.Duration) error {
```
Mala practica devolver solo `interface{}`, deberiamos usar genericos y devolver el tipo con el assertion ya hecha, asi solo el integrador tiene que usar el objecto, sin ningun tratamiento extra.

char2cs commented on internal/core/database/cache/gocache.go:
```diff
@@ -0,0 +1,94 @@
+}
+
+// Get retrieves a value from the cache.
+func (g *GoCache) Get(ctx context.Context, key string, dest interface{}) (bool, error) {
```
Mismo comment aca

char2cs commented on internal/core/database/cache/repository_cache.go:
```diff
@@ -0,0 +1,199 @@
+// RepositoryCache wraps a repository interface with caching functionality.
+// It implements the decorator pattern, transparently adding caching to all read operations
+// and invalidating cache on write operations.
+type RepositoryCache[T any] struct {
```
Aca estamos declarando el tipo y la assertion al mismo tiempo. Podriamos unir gocache y Repository cache para una mejor implementation, sin tanto scaffolding.

char2cs commented on internal/core/database/cache/memory_cache.go:
```diff
@@ -0,0 +1,136 @@
+
+// MemoryCache implements the Cache interface using an in-memory map.
+// This is a simple implementation that doesn't require external dependencies.
+type MemoryCache struct {
```
No creo que la vayamos a usar. Con la implementacion de `go-cache` es suficiente.

char2cs commented on internal/core/database/integration_test.go:
```diff
@@ -0,0 +1,181 @@
+}
+
+// Integration test with real database and cache
+func TestDatabaseWithCache_Integration(t *testing.T) {
```
Pudiendo hacer un `integration_test`, seria buena idea hacer tests que comparen la respuesta entre la implementacion con Cache y sin cache, para estar seguro que el modulo de database se comporta de la misma manera sin importar el wrapper.

char2cs commented on internal/core/database/cache/repository_cache.go:
```diff
@@ -0,0 +1,199 @@
+}
+
+// buildCacheKey creates a cache key for an entity ID.
+func (r *RepositoryCache[T]) buildEntityKey(id uuid.UUID) string {
```
Por que estan estos metodos aca? Si el repositorio actua sobre una entidad especifica, no deberiamos tener que hacer ningun prepend a las keys, solo guardar la key especifica del objeto.

valentin-villa replied:
Se utiliza para mejorar la legibilidad humana del caché.

Puede eliminarse sin afectar el funcionamiento, pero facilita el debugging y, dado que los cachés son un universo de diseño en sí mismos, lo considero una decisión válida.

char2cs commented on internal/core/database/cache/repository_cache.go:
```diff
@@ -0,0 +1,199 @@
+	}
+
+	// Cache the result
+	if err := r.cache.Set(ctx, key, entity, r.config.DefaultTTL); err != nil {
```
La funcion para cachear los resultados obtenidos puede ser asincronica usando `goroutines` al estilo `fire-and-forget`, asi no le pasamos el tiempo de espera para cachear al cliente (ojo este tiempo seria bajisimo, pero si estamos hablando de un EventSourcing que puede estar consumiendo muchas lecturas/escrituras, cada ms se agradece)

valentin-villa replied:
Genera problemas en testing.

Probé utilizando goroutines, pero los tests de integración no aseguran la paridad entre la base de datos y el caché.
Por como lo entiendo yo, esto nos obliga a elegir entre usar goroutines y aceptar tests poco confiables, o no usarlas y garantizar dicha paridad.

char2cs commented on internal/core/database/cache/repository_cache.go:
```diff
@@ -0,0 +1,199 @@
+	return created, nil
+}
+
+// extractID extracts the ID field from an entity using reflection.
```
`reflection` es un patron poco usando en go y muy poco performante. Si el codigo necesita `reflection`, habria que pensar en un refactor o volver a pensar esto. 

Veo que lo estas usando para extraer el ID. Esto puede ser mucho mas facil, por ahi tambien habria que cambiar la implementacion normal del `sqlite`, pero:
- Podriamos tener un parametro que nos devuelva la llave por la cual el integrador quiere guardar la entity, esto nos permite reaccionar de forma mas facil, cada vez que necesitemos la `key`, llamamos a ese metodo. Es responsabilidad del integrador decirnos cual es su key.
```go
Repository[T](
    name string,
    GetID func(entity T) string,
    opts Repository.Options? // no me acuerdo como estaba.
)
```
- El cache funcionaria como un wrapper, si esta encendido, en el constructor creamos el respositorio con cache, pasandole como parametro el `base`. 
- El cliente es agnostico a lo que pase adentro del repositorio, el solo lo configuro y espera poder guardarlo.

valentin-villa replied:
Esto es muy util y quiero eliminar el uso de reflection asi que tu comentario es oro.

Otra alternatiava es crear una interfaz que los objetos que vayamos a cachear satisfagan. Algo como:
```
type Identifiable interface {
  GetID() uuid.UUID
}
```

Mientras tanto, trato de implementar tu solucion

char2cs commented on internal/core/config/default.yaml:
```diff
@@ -27,3 +27,8 @@ config:
+  cache:
+    enabled: false
+    default_ttl: "5m"
+    cleanup_interval: "1m"
```
Estos son parametros que 100% pueden ir en el constructor del repository, no en la config general de Quiver.

char2cs reviewed (changes requested):
Sacando los comentarios, buena implementación y usás bien los conceptos de Go. Trataría de reducir la cantidad de archivos y aplicar mejor algunos conceptos del lenguaje, pero le estás agarrando la mano.

Valentin-Vi committed:
Removed default config from default.yaml and defined it as constant values in internal/core/database/cache/config.go

Valentin-Vi committed:
Removed unnecessary in-memory cache implementation

Valentin-Vi committed:
Removed unnecessary tests for removed in-memory cache feature

Valentin-Vi committed:
Added integration tests

Valentin-Vi committed:
Minor changes ;)

Valentin-Vi committed:
Minor changes ;)

Valentin-Vi committed:
Streamlined cached_repository.go

Valentin-Vi committed:
The cached repository now uses dependency injection for its base repository. Fixed tests. Removed unused field Context from builder.go.

Valentin-Vi committed:
Removed comments from tests

Valentin-Vi committed:
Added Where method in internal/core/database/cache/cached_repository.go to satisfy new RepositoryInterface

Valentin-Vi committed:
Took test coverage of internal/core/database/cache/cached_repository.go up to 94.4%. Fixed missing error handling and implemented use of Watcher.Warn for configuration errors and cache write failure

Valentin-Vi committed:
Removed Watcher logging at internal/core/database/cache/cached_repository.go constructor.

Valentin-Vi committed:
Merge branch 'develop' into feature/cache-abstraction-layer

char2cs commented on internal/core/database/cache/cached_repository.go:
```diff
@@ -0,0 +1,216 @@
+}
+
+func (cr *CachedRepository[T]) Create(ctx context.Context, entity *T) (*T, error) {
+	return cr.db.Create(ctx, entity)
```
Acá podemos aprovechar y cachear lo que va a la DB. Así, cuando lo queramos recuperar, ya está cacheado en memoria.

char2cs commented on internal/core/database/cache/cached_repository.go:
```diff
@@ -0,0 +1,216 @@
+		return nil, err
+	}
+
+	for _, entity := range result {
```
Aca valen guarda, estamos reimplementando la funcion del repositorio para guardar. 

La capa de cache deberia ser un wrapper por sobre el repositorio original:
```go
	v, found := cr.cache.GetById(key) // Primero buscamos en cache
	if found {
		data, ok := v.([]byte)
		if !ok {
			return nil, cacheerr.ErrInvalidCacheValue
		}
		var result []*T
		err := json.Unmarshal(data, &result)
		if err != nil {
			return nil, err
		}
		return result, nil  // Si esta devolvemos eso
	}

    entity, err := baseRepository.GetById() // Si no fallbackeamos al repositorio original
    if err != nil {
         return err
    }

    go cr.cache.AsyncSet(entity) // Asincronicamente hidratamos el cache

    return entity // Y devolvemos la respuesta
```

char2cs commented on internal/core/database/cache/cached_repository.go:
```diff
@@ -0,0 +1,216 @@
+		return uuid.Nil, false
+	}
+
+	val := reflect.ValueOf(entity)
```
Mucho cuidado con `reflect`, no es recomendable. 

Aca lo que podemos hacer es pedirle, en el constructor del cache, un metodo ExtractKey(entity T) string. Esto actuaria como una funcion que sirve para extraer la key de la entidad. Nos permite ser flexibles, y no usar `reflect` (parece magico, pero es muy poco performante)

char2cs commented on internal/core/database/builder.go:
```diff
@@ -0,0 +1,47 @@
+	}
+}
+
+func (b *DatabaseBuilder[T]) WithCache(config cache.CacheConfig) *DatabaseBuilder[T] {
```
Esto esta buenisimo. Como parte del otro comment, podriamos agregar aca un parametro mas:
```go
WithCache(config cache.CacheConfig, extractKey func(entity T) string)
```

Valentin-Vi committed:
Eliminated use of reflection, replacing it with a  param  which asks for a method/function to retrieve keys from a  (signature .

Valentin-Vi committed:
Replaced patrickmn/go-cache package for dgraph-io/ristretto package. Refactores cached repository, database builder, cache config, and adjacent tests. Removed Get method from repository interface and its implementations (cached repo and repo). Currently working to satisfy >90% test coverage.

Valentin-Vi committed:
Brought up test coverage percentage to >90.0% overall

char2cs commented on internal/core/database/cache/cached_repository.go:
```diff
@@ -0,0 +1,156 @@
+import (
+	"context"
+
+	"github.com/dgraph-io/ristretto/v2"
```
Me gusto este cambio de lib, parece que es mejor que cualquier otra en Go.

valentin-villa replied:
Fui medio lento en ponderar mi eleccion de lib pero si Ristretto esta muy buena y resolvio muchos prblemas de la implenentacion previa del cache.

char2cs commented on internal/core/database/builder.go:
```diff
@@ -0,0 +1,49 @@
+
+type DatabaseBuilder[T any] struct {
+	name        string
+	cacheConfig cache.CacheConfig
```
El tipo y struct de cache deberia estar dentro del paquete `core/database/cache`. 

No en `core/config`. El paquete de `core/config` son settings que dejamos que el usuario modifique, no aplica la config del cache. Sacaria ese struct de ahi y lo dejaria en este paquete.

char2cs commented on internal/core/config/config.go:
```diff
@@ -48,12 +48,19 @@ type Watcher struct {
 	Compress bool   `yaml:"compress"`
 }
 
+type Cache struct {
```
Sacaria este struct de aca.

char2cs reviewed (dismissed):
Todo lo demas lo veo bien, pero desacopemos el cache de la config global de Quiver. No aplica para este caso. 

Trabajo impecable Valen, cambiemos eso y tiene un approve.

Valentin-Vi committed:
Removed cache related configurations from . Renamed CacheConfig.DefaultTTL to CacheConfig.TTL and CacheConfig.DefaultCostOfItem to CacheConfig.CostOfItem.

Valentin-Vi committed:
Merge branch 'develop' into feature/cache-abstraction-layer

Valentin-Vi committed:
Added more tests for internal/core/database/repository/repository_test.go and internal/core/database/cache/cache_repository_test.go

Valentin-Vi committed:
Included invalid migration test to up test coverage of NewRepository method

Valentin-Vi committed:
Removed comments from tests. Fix: updated go.sum with missing transitive dependencies.

Valentin-Vi committed:
Removed troublesome and unnecessary test I had previously added few commits ago

Valentin-Vi committed:
Fix: Tests for Windows were failing because invalid path was valid on Windows.

Valentin-Vi committed:
Fix: Windows test keept failing because last dir name ("CON") was not invalid as of Windows 10. Changed it one again.
